src/unity/toolkits/object_detection/one_shot_object_detection/util/superposition.cpp

```diff
@@ -0,0 +1,128 @@
+/* Copyright © 2019 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ */
+#include <boost/gil/extension/numeric/sampler.hpp>
+#include <boost/gil/extension/numeric/resample.hpp>
+
+#include <unity/toolkits/object_detection/one_shot_object_detection/util/mapping_function.hpp>
+#include <unity/toolkits/object_detection/one_shot_object_detection/util/superposition.hpp>
+#include <unity/toolkits/object_detection/one_shot_object_detection/util/quadrilateral_geometry.hpp>
+
+#define BLACK boost::gil::rgb8_pixel_t(0,0,0)
+#define WHITE boost::gil::rgb8_pixel_t(255,255,255)
+#define RGBA_WHITE boost::gil::rgba8_pixel_t(255,255,255,0)
+
+namespace turi {
+namespace one_shot_object_detection {
+
+namespace data_augmentation {
+
+namespace superposition {
+
+void superimpose_image_rgb(const boost::gil::rgb8_image_t::view_t &masked,
+                           const boost::gil::rgb8_image_t::view_t &mask,
+                           const boost::gil::rgb8_image_t::view_t &transformed,
+                           const boost::gil::rgb8_image_t::view_t &mask_complement,
+                           const boost::gil::rgb8_image_t::view_t &background) {
+  for (int y = 0; y < masked.height(); ++y) {
+    auto masked_row_it = masked.row_begin(y);
+    auto mask_row_it = mask.row_begin(y);
+    auto transformed_row_it = transformed.row_begin(y);
+    auto mask_complement_row_it = mask_complement.row_begin(y);
+    auto background_row_it = background.row_begin(y);
+    for (int x = 0; x < masked.width(); ++x) {
+      masked_row_it[x][0] = (mask_row_it[x][0]/255 * transformed_row_it[x][0] + 
+        mask_complement_row_it[x][0]/255 * background_row_it[x][0]);
+      masked_row_it[x][1] = (mask_row_it[x][1]/255 * transformed_row_it[x][1] + 
+        mask_complement_row_it[x][1]/255 * background_row_it[x][1]);
+      masked_row_it[x][2] = (mask_row_it[x][2]/255 * transformed_row_it[x][2] + 
+        mask_complement_row_it[x][2]/255 * background_row_it[x][2]);
+    }
+  }
+}
+
+
+flex_image create_synthetic_rgb_image(const boost::gil::rgb8_image_t::view_t &background_view,
+                                      ParameterSampler &parameter_sampler,
+                                      const flex_image &object) {
+  Eigen::Matrix<float, 3, 3> M = parameter_sampler.get_transform().inverse();
+  boost::gil::rgb8_image_t::view_t starter_image_view = interleaved_view(
+    object.m_width,
+    object.m_height,
+    (boost::gil::rgb8_pixel_t*) (object.get_image_data()),
+    object.m_channels * object.m_width // row length in bytes
+  );
+  boost::gil::rgb8_image_t mask(boost::gil::rgb8_image_t::point_t(background_view.dimensions()));
```

Comment:
I think it's worth investigating a more time and space efficient data structure for the masks, since they're really just bitmaps and not entire images.

Reply:
Tracking this in #1859